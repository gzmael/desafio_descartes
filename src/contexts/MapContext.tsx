import { ReactNode, createContext, useCallback, useState } from 'react'
import { LatLngLiteral } from 'leaflet'
import { apiGoogleMaps } from 'lib/axios'

export interface MapContextProps {
  position: LatLngLiteral
  address: string
  isLoading: boolean
  updatePosition: (position: LatLngLiteral) => void
  getLocation: () => void
  getLocationByAddress: (address: string) => void
  updateAddress: (address: string) => void
}

export const MapContext = createContext({} as MapContextProps)

interface MapProviderProps {
  children: ReactNode
}

const MapProvider = ({ children }: MapProviderProps) => {
  const [position, setPosition] = useState<LatLngLiteral>({
    lat: -6.7931,
    lng: -39.307373,
  })
  const [address, setAddress] = useState('')
  const [isLoading, setIsLoading] = useState(false)

  const updatePosition = useCallback((position: LatLngLiteral) => {
    setPosition(() => position)
  }, [])

  const updateAddress = useCallback((address: string) => {
    setAddress(() => address)
  }, [])

  const getLocation = useCallback(() => {
    setIsLoading(true)
    navigator.geolocation.getCurrentPosition(
      (position) => {
        setPosition({
          lat: position.coords.latitude,
          lng: position.coords.longitude,
        })
      },
      (error) => {
        console.log(error)
      },
      {
        maximumAge: 1000,
        timeout: 3000,
        enableHighAccuracy: true,
      },
    )
    setIsLoading(false)
  }, [])

  const getLocationByAddress = useCallback(async (address: string) => {
    setIsLoading(true)

    const params = new URLSearchParams({ address })
    params.set('key', import.meta.env.VITE_GOOGLE_GEOCODING_KEY)

    try {
      const { data } = await apiGoogleMaps.get(`/json?${params.toString()}`)

      if (data && data?.status === 'OK') {
        const { lat, lng } = data.results[0].geometry.location
        setPosition({
          lat,
          lng,
        })
      }
    } catch (err) {
      console.log(err)
    } finally {
      setIsLoading(false)
    }
  }, [])

  return (
    <MapContext.Provider
      value={{
        position,
        address,
        isLoading,
        updatePosition,
        updateAddress,
        getLocation,
        getLocationByAddress,
      }}
    >
      {children}
    </MapContext.Provider>
  )
}

export { MapProvider }
