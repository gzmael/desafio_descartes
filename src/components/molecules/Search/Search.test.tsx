import {
  RenderOptions,
  fireEvent,
  render,
  screen,
} from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { Search } from '.'
import { vi } from 'vitest'
import { ReactElement } from 'react'
import { MapContext, MapContextProps } from 'contexts/MapContext'

const getLocation = vi.fn()
const getLocationByAddress = vi.fn((e: string) => console.log(e))
const updateAddress = vi.fn()
const updatePosition = vi.fn()

const props: MapContextProps = {
  address: '',
  getLocation,
  getLocationByAddress,
  isLoading: false,
  position: {
    lat: -6.7931,
    lng: -39.307373,
  },
  updateAddress,
  updatePosition,
  forecasts: [],
  backgroundUrl: '',
  weather: null,
}

const customRender = (
  ui: ReactElement,
  props: MapContextProps,
  options?: Omit<RenderOptions, 'wrapper'>,
) => {
  return render(
    <MapContext.Provider value={{ ...props }}>{ui}</MapContext.Provider>,
    { ...options },
  )
}

describe('Component/Molecules/Search', () => {
  it('should render Search Form with search and button', () => {
    const { container } = render(<Search />)

    expect(screen.getByRole('form')).toBeInTheDocument()
    const input = screen.getByRole('searchbox')

    expect(input).toBeInTheDocument()

    expect(container.firstChild).toHaveClass(
      'focus-within:ring-indigo-500focus-within:ring-1 flex w-full justify-between gap-1 rounded-lg bg-white px-2 focus-within:border-blue-200 focus-within:ring-2',
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should not be call props.onSearch when form is submitted and input is empty', async () => {
    render(<Search />)

    const form = screen.getByRole('form')

    fireEvent.submit(form)

    expect(getLocationByAddress).toHaveBeenCalledTimes(0)
  })

  it('should call props.onSearch when input search is valid', async () => {
    const inputText = 'VÃ¡rzea Alegre'

    customRender(<Search />, props)

    const input = screen.getByRole('searchbox')
    const form = screen.getByRole('form')

    expect(input).toBeInTheDocument()
    await userEvent.type(input, inputText)
    fireEvent.submit(form)
    expect(updateAddress).toHaveBeenLastCalledWith('e')
  })
})
