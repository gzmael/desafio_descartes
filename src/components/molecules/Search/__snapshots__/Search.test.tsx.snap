// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component/Molecules/Search > should render Search Form with search and button 1`] = `
<form
  class="focus-within:ring-indigo-500focus-within:ring-1 flex w-full justify-between gap-1 rounded-lg bg-white px-2 focus-within:border-blue-200 focus-within:ring-2"
  name="search-form"
>
  <input
    class="w-full bg-transparent text-sm text-blue-700 focus:outline-none"
    id="search"
    name="q"
    placeholder="Procure por um local"
    type="search"
    value=""
  />
  <button
    class="flex items-center justify-center font-medium disabled:cursor-not-allowed cursor-pointer disabled:opacity-50 disabled:pointer-events-none focus:outline-none active:scale-95 bg-transparent hover:bg-transparent transition-all focus:ring-none h-10 gap-2 text-base btn-md rounded w-auto p-2 text-light-blue-800"
    disabled=""
    title="Buscar"
    type="submit"
  >
    <svg
      class="lucide lucide-search"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="11"
        cy="11"
        r="8"
      />
      <line
        x1="21"
        x2="16.65"
        y1="21"
        y2="16.65"
      />
    </svg>
  </button>
</form>
`;
